Run alarm in background, select alarm file, alarm duration, and more. Also added common dev tooling

bbelderbos opened the pull request.

pmayd commented on alarm/__init__.py:
```diff
@@ -1 +1,3 @@
-__version__ = '0.1.0'
+from importlib.metadata import version
+
+__version__ = version(__name__)
```
just learned something new ;)

bbelderbos replied:
got it from Karmabot actually :)

bbelderbos replied:
I am changing this to the flit way:

```
"""A package to play an alarm after an interval"""

__version__ = "0.0.1"
```

then in toml:

```
[project]
...
dynamic = ["version", "description"]
``

pmayd commented on alarm/alarm.py:
```diff
@@ -1,58 +1,160 @@
-ALARM_MUSIC_FILE = config('ALARM_MUSIC_FILE')
-SECONDS_IN_MIN = 60
-POMODORO_TIMING = 25 * SECONDS_IN_MIN
+from . import __version__ as version
```
__version__ is really common in Python community for representing the version of the module so I would not gave it another alias like `version` that looks like a normal variable and thus the reader expects it to hold something else than the module's version, right?

bbelderbos replied:
ok we can use the dunder, happy with that, thanks

pmayd commented on alarm/alarm.py:
```diff
@@ -1,58 +1,160 @@
+
+    alarm_file_group = parser.add_mutually_exclusive_group()
+    alarm_file_group.add_argument(
+        "-l", "--song_library", help="Take a random song from a song library directory"
```
mention that you only support mp3/4?

bbelderbos replied:
good one, will add

pmayd commented on alarm/alarm.py:
```diff
@@ -1,58 +1,160 @@
+    elif args.file:
+        return args.file
+    else:
+        return os.environ["ALARM_MUSIC_FILE"]
```
this will fail if it is not set. Also, `args.file` could be an illegal choice if it points to a non existing file. Maybe we want to test all these and throw an AlarmFileException?

bbelderbos replied:
nice, done

pmayd commented on alarm/alarm.py:
```diff
@@ -1,58 +1,160 @@
+    """Make sure we get a music file that exists"""
+    if not Path(file).exists():
+        raise AlarmFileException(f"{file} does not exist")
+    allowed_extensions = ("mp3", "mp4", "wav")
```
`wav` is not part of the `rglob` lookup in the _get_file method

bbelderbos replied:
oops, there you go, the danger of duplicated code haha - will fix

bbelderbos replied:
done

pmayd commented on alarm/alarm.py:
```diff
@@ -1,58 +1,160 @@
+    if not Path(file).exists():
+        raise AlarmFileException(f"{file} does not exist")
+    allowed_extensions = ("mp3", "mp4", "wav")
+    if not Path(file).suffix.endswith(allowed_extensions):
```
why `endswith`? The suffix should already be the part behind the `.` so I suggest comparing the suffix directly with the allowed extensions, otherwise you would accept somethin like `.not_mp3`

bbelderbos replied:
must have missed `suffix` yep then it's a straight comparison

bbelderbos replied:
done

pmayd commented on alarm/alarm.py:
```diff
@@ -1,58 +1,160 @@
+    allowed_extensions = ("mp3", "mp4", "wav")
+    if not Path(file).suffix.endswith(allowed_extensions):
+        raise AlarmFileException(
+            f"{file} is not supported ({', '.join(allowed_extensions)} files are)"
```
Instead of "... (... are)", use two sentences: "OGG is not supported. MP3, MP4, ... are."

bbelderbos replied:
ok

bbelderbos replied:
done

pmayd commented on alarm/alarm.py:
```diff
@@ -1,58 +1,160 @@
+    return file
+
+
+def main(args=None):
```
what is the reason for `args` here? `main` is called without arguments at the end of the file. Is there a case that args will not be None? Only thing I can think of is `python -m alarm.py` but this only works with packages that have a __main__.py if I remember correctly.

bbelderbos replied:
I needed that for the tests

pmayd commented on alarm/alarm.py:
```diff
@@ -1,58 +1,160 @@
+
+
+def main(args=None):
+    args = args if args else parse_args(sys.argv[1:])
```
I normally put this in the `if __name__ == "__main__"` part and not inside the function but I guess this has to do with my previous question

bbelderbos replied:
I needed this for the tests, e.g.

```
    args = parse_args(["-s", "3", "-b", "-f", str(BIRDS_ALARM_FILE)])
    main(args)
```

pmayd commented on alarm/alarm.py:
```diff
@@ -1,58 +1,160 @@
+        package = __package__
+        module = Path(sys.argv[0]).stem
+
+        os.system(f"python -m {package}.{module} -s {seconds} -f '{alarm_file}' &")
```
Ah very nice, this explains my question about args I guess, but still this would be covered by args, right? But interesting that you can call a module via package.module directly without __main__.py?

bbelderbos replied:
yeah seems so and without __main__.py :)
thanks for that background process suggestion, makes this tool so much more valuable

pmayd commented:
I have a lot of problems with poetry so I would like to see a `setup.py` added to this package with a proper entry point definition so this would also make `alarm` directly available in the shell, both under Linux and Windows, without the need of some linux bash profile function magic :P

pmayd commented on alarm/alarm.py:
```diff
@@ -1,58 +1,160 @@
 
-from decouple import config
-from playsound import playsound  # requires PyObjC (Mac)
+from dotenv import load_dotenv
```
is dotenv missing as dependency from poetry?

pmayd replied:
its there sorry

pmayd commented:
Windows is really hard lol...playsound is not working under windows and throwing an error so I cannot test the library. However, I managed to create a simple setup.py that works for installing the package the normal way via `python setup.py install`:

```
from setuptools import setup, find_packages

setup(
    name="alarm",
    version="0.2.0",
    description="A useful module",
    author='"PyBites <support@pybit.es>"',
    author_email="support@pybit.es",
    packages=find_packages(),
    install_requires=[
        "playsound",
        "python-dotenv",
    ],  # external packages as dependencies
    entry_points={"console_scripts": ["alarm=alarm.alarm:main"]},
)
```

Using this, I had to change
```
load_dotenv()
```

to 
```
load_dotenv(Path().cwd())
```

otherwise load_dotenv throws an error that it cannot find the start path. Overall Python is really not a good portable language, even this very small script/cli makes a lot of problems!

bbelderbos commented:
> I have a lot of problems with poetry so I would like to see a `setup.py` added to this package with a proper entry point definition so this would also make `alarm` directly available in the shell, both under Linux and Windows, without the need of some linux bash profile function magic :P

will add a toml and use Flit ...

bbelderbos commented:
Damn wonder if we want to support Windows then at all ... what error did `playsound` throw?

bbelderbos commented:
@pmayd addressed all comments and switched to requirements and toml that should work with flit!

Assuming you have your virtual environment in `venv/` this works for me:

```
$ flit install --python venv/bin/python
$ which alarm
/Users/bbelderbos/code/pybites-alarm/venv/bin/alarm
$ alarm -v
alarm 0.0.1
```

Please approve the PR and I get it up on PyPI (adding workflow file now)

pmayd commented on requirements-dev.txt:
```diff
@@ -0,0 +1,4 @@
+pytest
```
dotenv is missing so on my system pytest fails because of it ;)

pmayd replied:
same with playsound (make install installs from requirements-dev.txt and these are missing)

bbelderbos replied:
sorry will add `-r requirements.txt` to `requirements-dev.txt`, missed that

pmayd commented:
On windows, running `pytest` works now,m but two tests are failing:

```
 pytest
================================================= test session starts =================================================
platform win32 -- Python 3.9.7, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
rootdir: D:\git\pybites-alarm
plugins: cov-2.12.1
collected 13 items

tests\test_alarm.py .F.F.........                                                                                [100%]

====================================================== FAILURES =======================================================
________________________________________________ test_play_alarm_file _________________________________________________

    def test_play_alarm_file():
        start = datetime.now()
        play_alarm_file(BIRDS_ALARM_FILE, 1)
        end = datetime.now()
>       assert int((end - start).seconds) == 1
E       assert 0 == 1
E        +  where 0 = int(0)
E        +    where 0 = (datetime.datetime(2021, 9, 23, 21, 12, 22, 770203) - datetime.datetime(2021, 9, 23, 21, 12, 22, 662391)).seconds

tests\test_alarm.py:35: AssertionError
------------------------------------------------ Captured stderr call -------------------------------------------------

    Error 305 for command:
        open "tests\payloads\birds.wav"
    Zus�tzliche Zeichen nach einer Zeichenkette mit Anf�hrungszeichen sind nicht erlaubt.

    Error 263 for command:
        close "tests\payloads\birds.wav"
    Ger�t ist nicht ge�ffnet oder wird vom MCI nicht erkannt.
Failed to close the file: "tests\payloads\birds.wav"
Process Process-1:
Traceback (most recent call last):
  File "C:\tools\miniconda3\envs\alarm\lib\multiprocessing\process.py", line 315, in _bootstrap
    self.run()
  File "C:\tools\miniconda3\envs\alarm\lib\multiprocessing\process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "C:\tools\miniconda3\envs\alarm\lib\site-packages\playsound.py", line 72, in _playsoundWin
    winCommand(u'open {}'.format(sound))
  File "C:\tools\miniconda3\envs\alarm\lib\site-packages\playsound.py", line 64, in winCommand
    raise PlaysoundException(exceptionMessage)
playsound.PlaysoundException:
    Error 305 for command:
        open "tests\payloads\birds.wav"
    Zus�tzliche Zeichen nach einer Zeichenkette mit Anf�hrungszeichen sind nicht erlaubt.
________________________________________ test_get_file_for_filled_mp3_library _________________________________________

tmpdir = local('C:\\Users\\micha\\AppData\\Local\\Temp\\pytest-of-micha\\pytest-0\\test_get_file_for_filled_mp3_l0')

    def test_get_file_for_filled_mp3_library(tmpdir):
        music_dir = tmpdir
        songs = {music_dir / f"{song}.mp3" for song in "live love train enjoy".split()}
        for song in songs:
            open(song, "a").close()
        args = parse_args(["-s", "3", "-l", str(music_dir)])
>       assert _get_file(args) in songs
E       AssertionError: assert 'C:\\Users\\micha\\AppData\\Local\\Temp\\pytest-of-micha\\pytest-0\\test_get_file_for_filled_mp3_l0\\train.mp3' in {local('C:\\Users\\micha\\AppData\\Local\\Temp\\pytest-of-micha\\pytest-0\\test_get_file_for_filled_mp3_l0\\enjoy.mp3'...local('C:\\Users\\micha\\AppData\\Local\\Temp\\pytest-of-micha\\pytest-0\\test_get_file_for_filled_mp3_l0\\train.mp3')}
E        +  where 'C:\\Users\\micha\\AppData\\Local\\Temp\\pytest-of-micha\\pytest-0\\test_get_file_for_filled_mp3_l0\\train.mp3' = _get_file(Namespace(seconds=3, minutes=None, background=False, display_timer=False, song_library='C:\\Users\\micha\\AppData\\Local\\Temp\\pytest-of-micha\\pytest-0\\test_get_file_for_filled_mp3_l0', file=None))

tests\test_alarm.py:51: AssertionError
=============================================== short test summary info ===============================================
FAILED tests/test_alarm.py::test_play_alarm_file - assert 0 == 1
FAILED tests/test_alarm.py::test_get_file_for_filled_mp3_library - AssertionError: assert 'C:\\Users\\micha\\AppData\...
```

pmayd commented:
Rest looks good

pmayd reviewed (approved):
On windows, we still have small problems. Installation only installs requirements from requirements-dev.txt not from requirements.txt and -dev is missing some core requirements. And two tests are failing.

bbelderbos commented:
> On windows, we still have small problems. Installation only installs requirements from requirements-dev.txt not from requirements.txt and -dev is missing some core requirements. And two tests are failing.

So one is related to `playsound`, does the alarm work at all on Windows? Let's see what the tests do on GitHub Actions upon merge ... Thanks for reviewing!

bbelderbos commented:
`load_dotenv(Path().cwd())` works for me, maybe we go with that change ...
What about `dotenv_values(".env")`, does that work for you? https://pybit.es/articles/how-to-handle-environment-variables-in-python/
